sympy/combinatorics/homomorphisms.py

```diff
@@ -103,28 +131,39 @@ def image(self):
             values = list(set(self.images.values()))
             if isinstance(self.codomain, PermutationGroup):
                 self._image = self.codomain.subgroup(values)
-            elif isinstance(self.codomain, FpGroup):
-                self._image = FpSubgroup(self.codomain, values)
             else:
-                self._image = FreeSubgroup(self.codomain, values)
+                self._image = FpSubgroup(self.codomain, values)
         return self._image
 
     def _apply(self, elem):
         '''
         Apply `self` to `elem`.
 
         '''
-        if not elem in self.domain.free_group:
+        if not elem in self.domain:
             raise ValueError("The supplied element doesn't belong to the domain")
         if elem.is_identity:
             return self.codomain.identity
         else:
-            p = elem.array_form[0][1]
-            if p < 0:
-                g = elem[0]**-1
+            images = self.images
+            value = self.codomain.identity
+            if isinstance(self.domain, PermutationGroup):
+                gens = self.domain.generator_product(elem, original=True)
+                for g in gens:
+                    if g in self.images:
+                        value = images[g]*value
```

Comment:
If `generator_product(elem)` returns `[s_1 ... s_n]`, `g == s_n*...*s_1`, and `value = images[g]*value` seems to be consistent with it.